File: one_edit_distance.py
# one edit distance

def one_edit_distance(s, t):
    ns, nt = len(s), len(t)

    # make s be the shorter var
    if ns > nt:
        return one_edit_distance(t, s)

    # if length difference is more than 1, dont bother
    if nt - ns > 1:
        return False

    # check chars one by one
    for i in range(ns):
        # here we've found a difference
        if s[i] != t[i]:
            if ns == nt: #both strigs are same size, so remaining characters must match
                return s[i+1:] == t[i+1:]
            else: # strings are different size, remaining chars from t+1 must match
                return s[i:] == t[i+1:]

    # no differences so far, last check is make sure both strings are not same size
    return ns != nt


a1 = "abxcdf"
b1 = "abxcff"

a2 = "abcde"
b2 = "acb"

a3 = "ac"
b3 = "acx"

one_edit_distance(a1, b1)

assert one_edit_distance(a1, b1) == True
assert one_edit_distance(a2, b2) == False
assert one_edit_distance(a3, b3) == True

File: arrays_and_strings/one_edit_distance.php
<?php
// one edit distance

class Solution
{
    public function one_edit_distance($str1, $str2) {

        //str1 must be shorter or equal
        if (strlen($str1) > strlen($str2)) {
            return $this->one_edit_distance($str2, $str1);
        }

        //too big of a difference?
        if (strlen($str2) - strlen($str1) > 1) {
            return false;
        }

        for ($i = 0; $i < strlen($str1); $i++) {
            if ($str1[$i] != $str2[$i]) {
                // a difference is found, remaining characters must be equal
                if (strlen($str1) == strlen($str2)) {
                    return substr($str1, $i+1) == substr($str2, $i+1);
                } else {
                    return substr($str1, $i) == substr($str2, $i+1);
                }
            } 
        }

        // both strings are equal
        return $str1 != $str2;
    }
}

$s = new Solution();

$str1 = "abababababahfhfp";
$str2 = "abababababahfhfs";

$ans = $s->one_edit_distance($str1, $str2);

var_dump($ans);

?>
